### PR TITLE
e2e_node: DRA: test plugin failures

### LINKED ISSUE #124608: DRA: add e2e_node tests
### What would you like to be added?

https://github.com/kubernetes/kubernetes/pull/124323 reminded us that test coverage of checkpointing and the interaction between kubelet and DRA drivers could be improved. We can do error injection and/or add delays by extending https://github.com/kubernetes/kubernetes/blob/4946c1fde24174ca71a9c122330f2fdf335593ea/test/e2e/dra/deploy.go#L420-L431 such that a test can add callbacks. This needs to be connected to the `NewDriver` call in E2E tests, perhaps by extending the `configureResources` call such that it both configures resources and the driver.

Here are some test scenarios that may be relevant. In all cases the test must ensure that NodePreparedResources succeeds before running a pod and NodeUnprepareResources succeeds for all claims which were previously prepared (not called out explicitly below). All pods must stop running.

## Retry NodePrepareResources

- single claim, single driver
- NodePrepareResources fails until allowed to succeed
- wait that NodePrepareResources failed at least once
- ensure that pod does not get started for a certain duration
- allow NodePrepareResources to succeed
- wait for pod to run
- kill pod

## Retry NodeUnprepareResources

- single claim, single driver
- run pod
- kill pod
- NodeUnprepareResources fails until allowed to succeed
- wait that NodeUnprepareResources failed at least once
- ensure that pod is not marked as completed for a certain duration
- allow NodeUnprepareResources to succeed

## Retry NodePrepareResources after restart

As above, but now also restart kubelet before allowing NodePrepareResources to succeed.

## Retry NodeUnprepareResources

Same for NodeUnprepareResources.

## Partial success for NodePrepareResources

- single claim, two drivers for that claim
- NodePrepareResources for driver A succeeds, the one for driver B fails.
- Rest of the test as for "NodePrepareResources".

## Partial success for NodeUnprepareResources

Same for NodeUnprepareResources.

## Partial success for NodePrepareResources with restart

Restart kubelet while the NodePrepareResources call for driver B is running.

## Partial success for NodeUnprepareResources

Restart kubelet while the NodeUnprepareResources call for driver B is running.

## Pod deletion during NodePrepareResources

- single claim, single driver
- stop kubelet while NodePrepareResources is running
- (force-)delete pod
- ensure that pod is truly gone
- restart kubelet
- ensure that NodePrepareResources does not get called again for a certain period of time

## Pod deletion during NodeUnprepareResources

- single claim, single driver
- stop kubelet while NodeUnprepareResources is running
- (force-)delete pod
- ensure that pod is truly gone
- restart kubelet
- ensure that NodeUnprepareResources gets called

/cc @klueska 
/assign @bart0sh 
/sig node

### Why is this needed?

Better test coverage.

### PR DESCRIPTION
**Pull request by bart0sh**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Test DRA plugin failures

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/124608

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

### CODE REVIEW

**Comment by bart0sh** [comment]
@klueska I've just pushed new test cases + the latest changes from https://github.com/kubernetes/kubernetes/pull/124323.

In my setup 2 test cases that test 2 Kubelet plugins fail, the rest work just fine:
```
sudo _output/local/go/bin/e2e_node.test --ginkgo.focus='DRA' --ginkgo.v --report-dir=/tmp/e2e_node.test/logs --node-name=devel --kubelet-flags='--cluster-domain=cluster.local --fail-swap-on=false --cgroup-driver=systemd' --prepull-images=false --container-runtime-endpoint=unix:///run/containerd/containerd.sock --kubelet-config-file=test/e2e_node/jenkins/default-kubelet-config.yaml --runtime-config=api/all=true --service-feature-gates=Dyn
amicResourceAllocation=true --feature-gates=DynamicResourceAllocation=true

...


```

I'll try to fix this today and add the rest of test cases described in the https://github.com/kubernetes/kubernetes/issues/124608

**Comment by bart0sh** [comment]
Status update: these new test cases implemented and passing:
- Retry NodePrepareResources
- Retry NodeUnprepareResources
- Retry NodePrepareResources after restart
- Retry NodeUnprepareResources after restart
- Partial success for NodePrepareResources
- Partial success for NodeUnprepareResources
- Pod deletion during NodePrepareResources
- Pod deletion during NodeUnprepareResources

These should still to be implemented and debugged:
 - Partial success for NodeUnprepareResources
 - Partial success for NodePrepareResources with restart
 - Partial success for NodeUnprepareResources with restart
 - Pod deletion during NodePrepareResources
 - Pod deletion during NodeUnprepareResources

File: pkg/kubelet/cm/dra/manager.go [hunk 1]
```diff
@@ -225,7 +225,7 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 			for _, resourceHandle := range claimInfo.ResourceHandles {
 				// If no DriverName is provided in the resourceHandle, we
 				// use the DriverName from the status
-				pluginName := claimInfo.DriverName
+				pluginName := resourceHandle.DriverName
```
Comment (pohly) [inline pkg/kubelet/cm/dra/manager.go:228]: You can remove the if check and update the comment. We require that resourceHandle.DriverName is set.

Comment (pohly) [reply pkg/kubelet/cm/dra/manager.go:228]: https://github.com/kubernetes/kubernetes/pull/124075 clarified that in the documentation. The validation has always checked it.

Comment (bart0sh) [reply pkg/kubelet/cm/dra/manager.go:228]: Thanks, will remove.

Comment (klueska) [reply pkg/kubelet/cm/dra/manager.go:228]: It probably makes more sense to remove it in https://github.com/kubernetes/kubernetes/pull/124323 than here.

Comment (bart0sh) [reply pkg/kubelet/cm/dra/manager.go:228]: Sure, I'll push all fixes into https://github.com/kubernetes/kubernetes/pull/124323 when I'm done with the e2e tests.

BTW, should we check if resourceHandle.DriverName is not empty or it's validated somewhere else?

**Comment by bart0sh** [comment]
I'm done adding new test cases. They're all passing:
```
sudo _output/local/go/bin/e2e_node.test --ginkgo.focus="must call NodeUnprepareResources again if it's in progress for one plugin when Kubelet restarts" --ginkgo.v --report-dir=/tmp/e2e_node.test/logs --node-name=devel --kubelet-flags='--cluster-domain=cluster.local --fail-swap-on=false --cgroup-driver=systemd' --prepull-images=false --container-runtime-endpoint=unix:///run/containerd/containerd.sock --kubelet-config-file=test/e2e_node/jenkins/default-kubelet-config.yaml --runtime-config=api/all=true --service-feature-gates=DynamicResourceAllocation=true --feature-gates=DynamicResourceAllocation=true
...
Ran 16 of 583 Specs in 507.896 seconds
SUCCESS! -- 16 Passed | 0 Failed | 0 Pending | 567 Skipped
PASS
```

This PR is ready for review now. @pohly @klueska PTAL

**Comment by bart0sh** [comment]
2reviewers: This PR depends on https://github.com/kubernetes/kubernetes/pull/124323, so some tests may fail until https://github.com/kubernetes/kubernetes/pull/124323 is merged.

**Comment by bart0sh** [comment]
@klueska All new e2e_node test cases added by this PR pass with the code from https://github.com/kubernetes/kubernetes/pull/124323

File: test/e2e_node/dra_test.go [hunk 2]
```diff
@@ -111,51 +114,387 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
...
+
+		ginkgo.It("must run pod if NodeUnprepareResources fails and then succeeds", func(ctx context.Context) {
+			pod := createTestObjects(ctx, f.ClientSet, getNodeName(ctx, f), f.Namespace.Name, "draclass", "external-claim", "drapod", 10, true, []string{driverName})
+
+			ginkgo.By("wait for pod to run")
+			err := e2epod.WaitForPodRunningInNamespace(ctx, f.ClientSet, pod)
+			framework.ExpectNoError(err)
+
+			ginkgo.By("set failure mode for the plugin")
+			kubeletPlugin.SetFailure(errors.New("Simulated failure"))
```
Comment (pohly) [inline test/e2e_node/dra_test.go:167]: What causes the pod to stop? If it stops by without waiting for anything, then this test is racy (pod might stop and NodeUnprepareResources be called before `SetFailure`).

Comment (pohly) [reply test/e2e_node/dra_test.go:167]: I'm stopping here with the review (no pun intended), but perhaps you can check yourself whether other tests have the same issue (if it is one)?

File: test/e2e/dra/test-driver/app/kubeletplugin.go [hunk 3]
```diff
@@ -162,10 +163,18 @@ func (ex *ExamplePlugin) IsRegistered() bool {
...
-	ex.block = true
+	ex.mutex.Lock()
+}
+
+func (ex *ExamplePlugin) Unblock() {
+	ex.mutex.Unlock()
+}
+
+func (ex *ExamplePlugin) SetFailure(err error) {
+	ex.failure = err
```
[the remark this answers is not in the document]
Comment (bart0sh) [reply test/e2e/dra/test-driver/app/kubeletplugin.go:178]: > t would be a useful exercise to run go test -race ./test/e2e for the DRA tests.

Can you explain how to enable "DynamicResourceAllocation" feature gate this way?

File: test/e2e/dra/test-driver/app/gomega.go [hunk 4]
```diff
@@ -33,22 +33,42 @@ var BeRegistered = gcustom.MakeMatcher(func(actualCalls []GRPCCall) (bool, error
 	return false, nil
 }).WithMessage("contain successful NotifyRegistrationStatus call")
 
-// NodePrepareResouceCalled checks that NodePrepareResource API has been called
-var NodePrepareResourceCalled = gcustom.MakeMatcher(func(actualCalls []GRPCCall) (bool, error) {
+// NodePrepareResoucesSucceeded checks that NodePrepareResources API has been called and succeeded
+var NodePrepareResourcesSucceeded = gcustom.MakeMatcher(func(actualCalls []GRPCCall) (bool, error) {
 	for _, call := range actualCalls {
-		if strings.HasSuffix(call.FullMethod, "/NodePrepareResource") && call.Err == nil {
+		if strings.HasSuffix(call.FullMethod, "/NodePrepareResources") && call.Err == nil {
```
[the remark this answers is not in the document]
Comment (bart0sh) [reply test/e2e/dra/test-driver/app/gomega.go]: Thank you for the review! added `&& call.Response != nil` as suggested.

File: test/e2e/dra/test-driver/app/kubeletplugin.go [hunk 5]
```diff
@@ -162,10 +163,18 @@ func (ex *ExamplePlugin) IsRegistered() bool {
 	return status.PluginRegistered
 }
 
-// Block sets a flag to block Node[Un]PrepareResources
+// Block locks plugin mutext to block Node[Un]PrepareResources
```
[the remark this answers is not in the document]
Comment (bart0sh) [reply test/e2e/dra/test-driver/app/kubeletplugin.go]: Correct. So far the only two calls that take it are `NodePrepareResources` and `NodeUnprepareresources`.

File: test/e2e_node/dra_test.go (hunk 2 above)
Comment (bart0sh) [reply test/e2e_node/dra_test.go:167]: > What causes the pod to stop? If it stops by without waiting for anything, then this test is racy (pod might stop and NodeUnprepareResources be called before SetFailure).

The pod sleeps for 10 secods. I hope this time is enough to set failure mode for UnprepareResource.

File: test/e2e/dra/test-driver/app/kubeletplugin.go [hunk 6]
```diff
@@ -309,6 +311,9 @@ func (ex *ExamplePlugin) NodePrepareResources(ctx context.Context, req *drapbv1a
 	resp := &drapbv1alpha3.NodePrepareResourcesResponse{
 		Claims: make(map[string]*drapbv1alpha3.NodePrepareResourceResponse),
 	}
+	if ex.failure != nil {
```
[the remark this answers is not in the document]
Comment (bart0sh) [reply test/e2e/dra/test-driver/app/kubeletplugin.go]: wrapped into mutex.Lock/Unlock, thanks.

File: test/e2e/dra/test-driver/app/kubeletplugin.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (bart0sh) [reply test/e2e/dra/test-driver/app/kubeletplugin.go]: Actually other calls take it as well, but it doesn't influence test cases it seems. I can add another mutex if you think it's safer not to use this one.

File: test/e2e_node/dra_test.go (hunk 2 above)
Comment (pohly) [reply test/e2e_node/dra_test.go:167]: On a loaded system I have seen failures that could only be explained by code not running for quite some time, longer than 10 seconds.

Instead of depending on timing, can these tests be written so that there is a causal relationship between the different steps?

File: test/e2e/dra/test-driver/app/kubeletplugin.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (pohly) [reply test/e2e/dra/test-driver/app/kubeletplugin.go:179]: `go test` has a 10 minute default timeout.

If you prefer to run under ginkgo, then use `ginkgo -race`.

File: test/e2e_node/dra_test.go (hunk 2 above)
Comment (bart0sh) [reply test/e2e_node/dra_test.go:167]: Not sure there is an issue there, but removed waiting for the pod running condition. PTAL

Comment (bart0sh) [reply test/e2e_node/dra_test.go:167]: If that's not enough, I'll consider introducing separate locks for Prepare* And Unprepare* calls. Just let me know if you want that.

File: test/e2e/dra/test-driver/app/kubeletplugin.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bart0sh) [reply test/e2e/dra/test-driver/app/kubeletplugin.go:180]: running e2e_node.test with --ginkgo-flags='-race' didn't trigger any failures/panics.

File: test/e2e/dra/test-driver/app/kubeletplugin.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (pohly) [reply test/e2e/dra/test-driver/app/kubeletplugin.go]: Let's add a separate mutex.

The REST proxy (if merged) might need to lock the shared mutex and that shouldn't get blocked.

File: test/e2e_node/dra_test.go (hunk 2 above)
Comment (pohly) [reply test/e2e_node/dra_test.go:166]: This test still has a race.

What I am looking for is a guarantee that this sequence of events occurs regardless of timing:
- pod starts
-  NodePrepareResources succeeded
- pod stops
- NodeUnprepareResources fails at least once
- NodeUnprepareResources succeeds
- pod shutdown completes

Making `kubeletPlugin.SetFailure` specific to certain methods would ensure that. Then the failure mode can be enabled for NodeUnprepareResources before creating the pod.

One nice side effect is that the pod can be allowed to run to completion immediately (no need for a 10 second delay). The test then completes more quickly.

File: test/e2e_node/dra_test.go [hunk 7]
```diff
@@ -170,13 +497,13 @@ func newKubeletPlugin(ctx context.Context, nodeName string) *testdriver.ExampleP
 // NOTE: as scheduler and controller manager are not running by the Node e2e,
 // the objects must contain all required data to be processed correctly by the API server
 // and placed on the node without involving the scheduler and the DRA controller
-func createTestObjects(ctx context.Context, clientSet kubernetes.Interface, nodename, namespace, className, claimName, podName string) *v1.Pod {
+func createTestObjects(ctx context.Context, clientSet kubernetes.Interface, nodename, namespace, className, claimName, podName string, sleepTime int, deferPodDeletion bool, pluginNames []string) *v1.Pod {
```
Comment (pohly) [inline test/e2e_node/dra_test.go:500]: Any test that needs `sleepTime > 0` is suspicious. Can you rewrite all tests which do this and remove the parameter?

File: test/e2e_node/dra_test.go [hunk 8]
```diff
@@ -231,22 +558,26 @@ func createTestObjects(ctx context.Context, clientSet kubernetes.Interface, node
 	createdPod, err := clientSet.CoreV1().Pods(namespace).Create(ctx, pod, metav1.CreateOptions{})
 	framework.ExpectNoError(err)
 
-	ginkgo.DeferCleanup(clientSet.CoreV1().Pods(namespace).Delete, podName, metav1.DeleteOptions{})
+	if deferPodDeletion {
+		ginkgo.DeferCleanup(clientSet.CoreV1().Pods(namespace).Delete, podName, metav1.DeleteOptions{})
```
Comment (pohly) [inline test/e2e_node/dra_test.go:562]: You probably added `deferPodDeletion` because some test now needs to delete the pod and then this call failed.

You can remove the parameter, always do the DeferCleanup, and ignore "not found" errors:
```suggestion
		ginkgo.DeferCleanup(framework.IgnoreNotFound(clientSet.CoreV1().Pods(namespace).Delete), podName, metav1.DeleteOptions{})
```

https://github.com/kubernetes/kubernetes/blob/8c1983ffc0b6fe2293fc721cef8d961d79aafc53/test/e2e/framework/ginkgowrapper.go#L91-L96

**Review by pohly — CHANGES_REQUESTED** [review]

File: test/e2e_node/dra_test.go (hunk 8 above)
Comment (bart0sh) [reply test/e2e_node/dra_test.go:562]: > You can remove the parameter, always do the DeferCleanup, and ignore "not found" errors:

Have I understood correctly that you're proposing to ignore pod deletion for all test cases, most of which don't delete a pod?

File: test/e2e_node/dra_test.go (hunk 7 above)
Comment (bart0sh) [reply test/e2e_node/dra_test.go]: done

File: test/e2e_node/dra_test.go (hunk 2 above)
Comment (bart0sh) [reply test/e2e_node/dra_test.go]: done, PTAL

File: test/e2e_node/dra_test.go (hunk 8 above)
Comment (pohly) [reply test/e2e_node/dra_test.go:538]: Correct. This cleanup code only needs to ensure that the pod gets deleted. If that has been done already, then that's not a test failure.

Comment (bart0sh) [reply test/e2e_node/dra_test.go:538]: In most of the tests it still makes sense to check if pod is deleted, at least to my understanding. Pod is deleted in a couple of test cases only. Am I missing something?

Comment (pohly) [reply test/e2e_node/dra_test.go:538]: You always delete here. Then if the error is "NotFound", ignore the error because the desired state has already been reached.

Comment (bart0sh) [reply test/e2e_node/dra_test.go:538]: Ignoring the fact that pod doesn't exist when it should exist is worse than having a parameter in my opinion. WDYT?

Comment (bart0sh) [reply test/e2e_node/dra_test.go:538]: > You always delete here. Then if the error is "NotFound", ignore the error because the desired state has already been reached.

Not sure I follow. In most of test cases we shouldn't ignore "NotFound" error as the pod should exist. If it doesn't exist, then cleanup should fail.

Comment (pohly) [reply test/e2e_node/dra_test.go:538]: We don't need to be that strict. In other words, "unexpected pod deletion" is not something that we need to test for.

Comment (bart0sh) [reply test/e2e_node/dra_test.go:538]: I don't agree. We need to test for it explicitly, but we shouldn't regress tests quality just because we don't want to add one parameter.

File: test/e2e/dra/test-driver/app/kubeletplugin.go [hunk 9]
```diff
@@ -162,10 +171,62 @@ func (ex *ExamplePlugin) IsRegistered() bool {
...
+
+func (ex *ExamplePlugin) withLock(mutex *sync.Mutex, f func()) {
+	mutex.Lock()
+	f()
+	mutex.Unlock()
+}
+
+func (ex *ExamplePlugin) WithNodePrepareResourcesFailure(f func() (*v1.Pod, func())) (*v1.Pod, func()) {
+	ex.withLock(&ex.failPrepareResourcesMutex, func() {
+		ex.prepareResourcesFailure = errors.New("simulated PrepareResources failure")
```
[the remark this answers is not in the document]
Comment (bart0sh) [reply test/e2e/dra/test-driver/app/kubeletplugin.go]: done, PTAL

File: test/e2e/dra/test-driver/app/kubeletplugin.go [hunk 10]
```diff
@@ -162,10 +171,60 @@ func (ex *ExamplePlugin) IsRegistered() bool {
...
 // to emulate time consuming or stuck calls
-func (ex *ExamplePlugin) Block() {
-	ex.block = true
+func (ex *ExamplePlugin) WithBlockedNodeUnprepareResources(f func() (*v1.Pod, func())) (*v1.Pod, func()) {
+	ex.blockUnprepareResourcesMutex.Lock()
+	defer ex.blockUnprepareResourcesMutex.Unlock()
+	return f()
+}
+
+func (ex *ExamplePlugin) WithNodePrepareResourcesFailure(f func() (*v1.Pod, func())) (*v1.Pod, func()) {
```
Comment (pohly) [inline test/e2e/dra/test-driver/app/kubeletplugin.go:190]: Documentation?

But see below first, perhaps we don't need this in this form.

File: test/e2e/dra/test-driver/app/kubeletplugin.go [hunk 11]
```diff
@@ -162,10 +171,60 @@ func (ex *ExamplePlugin) IsRegistered() bool {
...
+	defer func() {
+		ex.failPrepareResourcesMutex.Lock()
+		ex.prepareResourcesFailure = nil
+		ex.failPrepareResourcesMutex.Unlock()
+	}()
+
+	return f()
+}
+
+func (ex *ExamplePlugin) GetPrepareResourcesFailure() error {
```
Comment (pohly) [inline test/e2e/dra/test-driver/app/kubeletplugin.go:204]: This doesn't need to be exported? If it did, it should have been documented.

File: test/e2e/dra/test-driver/app/kubeletplugin.go [hunk 12]
```diff
@@ -162,10 +171,60 @@ func (ex *ExamplePlugin) IsRegistered() bool {
...
+	ex.failPrepareResourcesMutex.Lock()
+	defer ex.failPrepareResourcesMutex.Unlock()
+	return ex.prepareResourcesFailure
+}
+
+func (ex *ExamplePlugin) WithNodeUnprepareResourcesFailure(f func() (*v1.Pod, func())) (*v1.Pod, func()) {
```
Comment (pohly) [inline test/e2e/dra/test-driver/app/kubeletplugin.go:210]: Documentation?

File: test/e2e/dra/test-driver/app/kubeletplugin.go [hunk 13]
```diff
@@ -162,10 +171,60 @@ func (ex *ExamplePlugin) IsRegistered() bool {
...
+	defer func() {
+		ex.failUnprepareResourcesMutex.Lock()
+		ex.unprepareResourcesFailure = nil
+		ex.failUnprepareResourcesMutex.Unlock()
+	}()
+
+	return f()
+}
+
+func (ex *ExamplePlugin) GetUnprepareResourcesFailure() error {
```
Comment (pohly) [inline test/e2e/dra/test-driver/app/kubeletplugin.go:224]: Doesn't need to be exported?

File: test/e2e/dra/test-driver/app/kubeletplugin.go [hunk 14]
```diff
@@ -378,9 +430,17 @@ func (ex *ExamplePlugin) nodeUnprepareResource(ctx context.Context, claimName st
 }
 
 func (ex *ExamplePlugin) NodeUnprepareResources(ctx context.Context, req *drapbv1alpha3.NodeUnprepareResourcesRequest) (*drapbv1alpha3.NodeUnprepareResourcesResponse, error) {
+	ex.blockUnprepareResourcesMutex.Lock()
+	defer ex.blockUnprepareResourcesMutex.Unlock()
```
Comment (pohly) [inline test/e2e/dra/test-driver/app/kubeletplugin.go:434]: Is there a reason why this is blocking differently than NodePrepareResources?

In NodePrepareResources, it's the individual `nodePrepareResource` which blocks.

File: test/e2e/dra/test-driver/app/kubeletplugin.go [hunk 15]
```diff
@@ -487,3 +547,13 @@ func (ex *ExamplePlugin) GetGRPCCalls() []GRPCCall {
 	calls = append(calls, ex.gRPCCalls...)
 	return calls
 }
+
+func (ex *ExamplePlugin) CountCalls(methodSuffix string) int {
```
Comment (pohly) [inline test/e2e/dra/test-driver/app/kubeletplugin.go:551]: Documentation?

File: test/e2e/dra/test-driver/app/kubeletplugin.go [hunk 16]
```diff
@@ -162,10 +171,60 @@ func (ex *ExamplePlugin) IsRegistered() bool {
 	return status.PluginRegistered
 }
 
-// Block sets a flag to block Node[Un]PrepareResources
+// WithBlockedNodePrepareResources locks blockPrepareResourcesMutex to block NodePrepareResources
+// to emulate time consuming or stuck calls
```
Comment (pohly) [inline test/e2e/dra/test-driver/app/kubeletplugin.go:175]: This doesn't mention why there is a function parameter.

File: test/e2e_node/dra_test.go [hunk 17]
```diff
@@ -111,51 +113,352 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
 		})
 
 		ginkgo.It("must keep pod in pending state if NodePrepareResources times out", func(ctx context.Context) {
-			ginkgo.By("set delay for the NodePrepareResources call")
-			kubeletPlugin.Block()
-			pod := createTestObjects(ctx, f.ClientSet, getNodeName(ctx, f), f.Namespace.Name, "draclass", "external-claim", "drapod")
+			kubeletPlugin.WithBlockedNodePrepareResources(func() (*v1.Pod, func()) {
```
Comment (pohly) [inline test/e2e_node/dra_test.go:116]: I find it hard to follow the flow of the test because of this callback approach. Can we do it without those `With*` calls?

This could read like this instead:
```
unblock := kubeletPlugin.BlockNodePrepareResources()
defer unblock()

// code which runs while blocked
```

I'm aware that this puts a little more responsibility on the caller. But I think the additional clarity is worth it.

File: test/e2e_node/dra_test.go [hunk 18]
```diff
@@ -111,51 +113,352 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
...
+					"Pod should be in Pending state as resource preparation time outed")
+				return pod, nil
+			})
+		})
 
-			ginkgo.By("wait for pod to be in Pending state")
-			err := e2epod.WaitForPodCondition(ctx, f.ClientSet, f.Namespace.Name, pod.Name, "Pending", framework.PodStartShortTimeout, func(pod *v1.Pod) (bool, error) {
-				return pod.Status.Phase == v1.PodPending, nil
+		ginkgo.It("must run pod if NodePrepareResources fails and then succeeds", func(ctx context.Context) {
+			pod, _ := kubeletPlugin.WithNodePrepareResourcesFailure(func() (*v1.Pod, func()) {
```
Comment (pohly) [inline test/e2e_node/dra_test.go:135]: A function returning a function which then isn't used at all? This is an artifact that somewhere else this was needed, but it just looks very odd here. This shows that these `With*` calls aren't the right abstraction.

**Review by pohly — CHANGES_REQUESTED** [review]

File: test/e2e_node/dra_test.go (hunk 17 above)
Comment (bart0sh) [reply test/e2e_node/dra_test.go:116]: With the current approach it's more readable as it's easier to see the code that runs when `[Un]PrepareResources` is blocked. It adds a nice structure to the code from my point of view. Not sure I want to go back to the plain less unstructured code. Will do only if you insist. Please, confirm.

File: test/e2e_node/dra_test.go (hunk 18 above)
Comment (bart0sh) [reply test/e2e_node/dra_test.go:135]: It's used in other test case, I believe. It's a function to start the Kubelet again and it's simply not needed in this test case.

File: test/e2e_node/dra_test.go (hunk 17 above)
Comment (pohly) [reply test/e2e_node/dra_test.go:116]: Perhaps you as the author understood this. I as reviewer didn't not find it readable.

Please change it.

Comment (bart0sh) [reply test/e2e_node/dra_test.go]: changed.

File: test/e2e_node/dra_test.go (hunk 18 above)
Comment (bart0sh) [reply test/e2e_node/dra_test.go]: replaced the same way as WithBlocked*

File: test/e2e/dra/test-driver/app/kubeletplugin.go (hunk 15 above)
Comment (bart0sh) [reply test/e2e/dra/test-driver/app/kubeletplugin.go:551]: added

File: test/e2e/dra/test-driver/app/kubeletplugin.go (hunk 14 above)
Comment (bart0sh) [reply test/e2e/dra/test-driver/app/kubeletplugin.go]: done

File: test/e2e/dra/test-driver/app/kubeletplugin.go (hunk 13 above)
Comment (bart0sh) [reply test/e2e/dra/test-driver/app/kubeletplugin.go]: done

File: test/e2e/dra/test-driver/app/kubeletplugin.go (hunk 11 above)
Comment (bart0sh) [reply test/e2e/dra/test-driver/app/kubeletplugin.go]: done

**Comment by bart0sh** [comment]
@pohly I've addressed all your comments, please review again, thanks.

File: test/e2e_node/dra_test.go [hunk 19]
```diff
@@ -111,51 +113,342 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
...
+			ginkgo.By("wait for pod to be in Pending state")
+			err := e2epod.WaitForPodCondition(ctx, f.ClientSet, f.Namespace.Name, pod.Name, "Pending", framework.PodStartShortTimeout, func(pod *v1.Pod) (bool, error) {
+				return pod.Status.Phase == v1.PodPending, nil
+			})
+			framework.ExpectNoError(err)
+
+			ginkgo.By("wait for plugin2 NodePrepareResources call to fail")
+			gomega.Eventually(kubeletPlugin2.GetGRPCCalls).WithTimeout(draplugin.PluginClientTimeout * 2).Should(testdriver.NodePrepareResourcesFailed)
+
+			unset()
```
Comment (pohly) [inline test/e2e_node/dra_test.go:347]: `unset` what? A more descriptive name would have made this test more readable.

But we've discussed this too long already. I'll probably just merge.

File: test/e2e_node/dra_test.go (hunk 8 above)
Comment (pohly) [reply test/e2e_node/dra_test.go:528]: IMHO the `deferPodDeletion` should be the opposite: always delete to be on the safe side, but only treat "not found" in some tests.

But have it your way.

File: test/e2e_node/dra_test.go [hunk 20]
```diff
@@ -111,51 +113,342 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
...
 	framework.ExpectNoError(err, "create CDI directory")
+	endpoint := fmt.Sprintf(endpointTemplate, pluginName)
 	err = os.MkdirAll(filepath.Dir(endpoint), 0750)
 	framework.ExpectNoError(err, "create socket directory")
 
 	plugin, err := testdriver.StartPlugin(
 		ctx,
 		cdiDir,
-		driverName,
+		pluginName,
```
Comment (pohly) [inline test/e2e_node/dra_test.go:446]: Nit: we don't really have "plugin names" in DRA. The KEP only talks about "driver names" and "DRA drivers having kubelet plugins which register with that driver name".

**Review by pohly — COMMENTED** [review]
/lgtm
/approve

**Comment by bart0sh** [comment]
@pohly Thank you for reviewing and merging this PR! I'll address your remaining comments in a separate PR.
